8356372: JVMTI heap sampling not working properly with outside TLAB allocations

stefank opened the pull request:
While working on improving the TLAB sizing code for ZGC @kstefanj ran into an issue with the following tests failing:
serviceability/jvmti/HeapMonitor/MyPackage/HeapMonitorInterpreterObjectTest.java
serviceability/jvmti/HeapMonitor/MyPackage/HeapMonitorStatObjectCorrectnessTest.java

The reason for seeing the problems now is that with the new sizing code ZGC used smaller TLABs at first, before resizing them to a proper size (to lower the waste). In the HeapMonitor tests we don't allocate enough to trigger GCs that will resize the TLABs so most of the tests will now run with small TLABs. This should not be a problem, but it turns out the current sampling code is not working properly when you get a lot of outside TLAB allocations. You get those when trying to allocate a fairly large object (~1400B) that won't fit into the TLAB, but there are still quite a bit of room in the TLAB so we don't want to throw it away and take a new one.

The problem in the current code is that we keep track of when to sample with multiple variables and when getting out of TLAB allocations these get out of sync.

The proposed patch is the result of a restructuring and fixing of the the code that me and @kstefanj did to fix this issue.

The idea is to better account how much we have allocated in three different buckets:
* Outside of TLAB allocations
* Accounted TLAB allocations
* The last bit of TLAB allocations that hasn't been accounted yet

And then use the sum of that to compare against a *non-changing* threshold to see if it is time to take a sample.

There are a few things to think about when reading this code:
* The thread can allocate and retire multiple TLABs before we cross the sample threshold.
* The sampling can take multiple samples in a single TLAB
* Any overshooting of the sample threshold triggers only one sample and the extra bytes are ignored when checking for the next sample.

There are some restructuring in the patch to confine the ThreadHeapSampler variables and code. For example:

1) Moved accounting variables out of TLAB and into the ThreadHeapSampler
2) Moved thread allocation accounting and sampling code out of the TLAB
3) Moved retiring of TLABs out of make_parseable (needed to support (2))

Some of that could be extracted into a separate PR if that's deemed worthwhile.

Tested with the HeapMonitor tests various TLAB sizes.

If there are anyone using these APIs it would be nice to get feedback if these changes work well for you.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8356372](https://bugs.openjdk.org/browse/JDK-8356372): JVMTI heap sampling not working properly with outside TLAB allocations (**Bug** - P3)


### Reviewers
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Contributors
 * Stefan Johansson `<sjohanss@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25114/head:pull/25114` \
`$ git checkout pull/25114`

Update a local copy of the PR: \
`$ git checkout pull/25114` \
`$ git pull https://git.openjdk.org/jdk.git pull/25114/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25114`

View PR using the GUI difftool: \
`$ git pr show -t 25114`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25114.diff">https://git.openjdk.org/jdk/pull/25114.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25114#issuecomment-2862493678)
</details>

kstefanj commented on src/hotspot/share/runtime/threadHeapSampler.hpp:
```diff
@@ -53,9 +89,42 @@ class ThreadHeapSampler {
+    _tlab_top_at_sample_start = tlab_top;
+  }
+
+  void reset_after_sampling(HeapWord* tlab_top) {
```
```suggestion
  void reset_after_sample(HeapWord* tlab_top) {
```

kstefanj commented on src/hotspot/share/runtime/threadHeapSampler.cpp:
```diff
@@ -408,26 +409,35 @@ void ThreadHeapSampler::pick_next_sample(size_t overflowed_bytes) {
-  pick_next_sample(overflow_bytes);
+  pick_next_sample();
+
+  reset_after_sampling(tlab_top);
```
```suggestion
  reset_after_sample(tlab_top);
```

kstefanj reviewed (approved):
Looks great :)

Suggestion to name it `reset_after_sample(...)` instead of `reset_after_sampling(...)` to be more consistent.

stefank reviewed (commented):
Thanks for reviewing and pair-fixing!

sspitsyn commented on src/hotspot/share/runtime/threadHeapSampler.hpp:
```diff
@@ -53,9 +89,42 @@ class ThreadHeapSampler {
     pick_next_sample();
```
The old identifier is used: _bytes_until_sample.

sspitsyn replied:
This comment still need a fix as this name was removed by this fix.

sspitsyn commented on src/hotspot/share/runtime/threadHeapSampler.cpp:
```diff
@@ -396,10 +396,11 @@ void ThreadHeapSampler::pick_next_geometric_sample() {
   assert(result > 0 && result < static_cast<double>(SIZE_MAX), "Result is not in an acceptable range.");
   size_t interval = static_cast<size_t>(result);
-  _bytes_until_sample = interval;
+  _sample_threshold = interval;
```
Nit: The line 399 is not needed as the value is reset at 400.

stefank replied:
Nice catch. This is debugging code from the early stage of this patch. It set up a consistent state to make it easier to get consistent data from the test runs. This also means that we need to respin our testing to make sure that the randomization doesn't trigger anything that we didn't catch with the fixed _sample_threshold.

stefank replied:
The unit tests still pass after the fix for this.

sspitsyn commented on src/hotspot/share/runtime/threadHeapSampler.hpp:
```diff
@@ -27,23 +27,59 @@
+  ThreadHeapSampler() :
+      _tlab_top_at_sample_start(nullptr),
+      _accumulated_tlab_bytes_since_sample(0),
+      _accumulated_outside_tlab_bytes_since_sample(0) {
```
Nit: Would it make sense to initialize `_sample_threshold` as well to make it consistent?

stefank replied:
To me it makes sense. I didn't do it here because it gets initialized further down and I didn't want reviewers to get hung up on that change. But I'll take this as a go-ahead to make that change.

stefank commented on src/hotspot/share/runtime/threadHeapSampler.hpp:
```diff
@@ -53,9 +89,42 @@ class ThreadHeapSampler {
   }
 
-  size_t bytes_until_sample()                    { return _bytes_until_sample;   }
+  size_t bytes_since_last_sample(HeapWord* tlab_top) const {
```
I had it like that but previous reviewers complained about the name. I made it longer to make sure that it is as clear as possible what this value is about. IMHO, one of the reasons why this code was hard to figure out why it was broken was because the names where not precise enough to explain what was stored in the variables.

Let me try to go the other way and completely remove the "unsampled" name. And then you can take a look and see if you really hate it and if you do I'll take a stab at finding some shorter names that still are descriptive for the casual reader.

sspitsyn replied:
Okay, thanks.

sspitsyn commented:
Stefan, thank you for the update. The whole fix looks good in general. I like the refactoring.
One comment still need an update though as it has a removed/renamed field name.

stefank commented:
Thanks, @sspitsyn! I think I've addressed all your comments now except making the names shorter. I hope that's OK and if we really want to shorten then, I don't mind doing it in a follow-up RFE.

stefank commented:
/contributor add @kstefanj

sspitsyn reviewed (approved):
Looks good. Thank you for the update!
I'm okay with current naming. Consistency is more important than shortness.

kstefanj commented on src/hotspot/share/runtime/threadHeapSampler.cpp:
```diff
@@ -408,26 +408,35 @@ void ThreadHeapSampler::pick_next_sample(size_t overflowed_bytes) {
+  LogTarget(Debug, gc, tlab) log;
+  if (log.is_enabled()) {
+    const bool should_sample = bytes_since_last_sample(tlab_top) >= _sample_threshold;
+    log_debug(gc, tlab)("Should sample: %s _sample_threshold: %zu total: %zu tlab: %zu current tlab: %zu outside_tlab: %zu",
```
```suggestion
    log_debug(gc, tlab)("Should sample: %s _sample_threshold: %zu total: %zu tlab: %zu current tlab: %zu outside tlab: %zu",
```

stefank reviewed (commented):
Thanks for the reviews! If final testing passes I plan to integrate tomorrow (unless new reviewers show up).

stefank commented:
/integrate
